### PR TITLE
Add log filtering processor distro

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Added an experimental feature for logs emitted within an active tracing context to follow the Activity's sampling decision. The feature can be enabled by setting `OTEL_DOTNET_AZURE_MONITOR_EXPERIMENTAL_ENABLE_LOG_SAMPLING` environment variable to `true`.

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Azure.Monitor.OpenTelemetry.AspNetCore.csproj [hunk 1]
```diff
@@ -23,10 +23,10 @@
     <!-- Depending on monthly deliverables, we may switch between PackageReference or ProjectReference. Keeping both here to make the switch easier. -->
 
     <!-- FOR PUBLIC RELEASES, MUST USE PackageReference. THIS REQUIRES A STAGGERED RELEASE IF SHIPPING A NEW EXPORTER. -->
-    <PackageReference Include="Azure.Monitor.OpenTelemetry.Exporter" />
+    <!--<PackageReference Include="Azure.Monitor.OpenTelemetry.Exporter" />-->
 
     <!-- FOR LOCAL DEV, ProjectReference IS PREFERRED. -->
-    <!--<ProjectReference Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Azure.Monitor.OpenTelemetry.Exporter.csproj" />-->
+    <ProjectReference Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Azure.Monitor.OpenTelemetry.Exporter.csproj" />
   </ItemGroup>
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Azure.Monitor.OpenTelemetry.AspNetCore.csproj:30]: I'm going to kindly ask you to revert this change before merging. :)
New beta will be released tomorrow, can wait till then to revert.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/AzureMonitorAspNetCoreEventSource.cs [hunk 2]
```diff
@@ -51,6 +51,15 @@ public void ConfigureFailed(Exception ex)
...
 
+        [NonEvent]
+        public void GetEnvironmentVariableFailed(string envVarName, Exception ex)
+        {
+            if (IsEnabled(EventLevel.Error))
+            {
+                GetEnvironmentVariableFailed(envVarName, ex.FlattenException().ToInvariantString());
+            }
+        }
+
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/AzureMonitorAspNetCoreEventSource.cs:62]: Please move this adjacent to your other method (just above 75). 
That makes it easier to read these.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/LogFilteringProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    internal class LogFilteringProcessor : BatchLogRecordExportProcessor
+    {
+        public LogFilteringProcessor(BaseExporter<LogRecord> exporter)
+            : base(exporter)
+        {
+        }
+
+        public override void OnEnd(LogRecord logRecord)
+        {
+            if (logRecord.SpanId == default || logRecord.TraceFlags == ActivityTraceFlags.Recorded)
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/LogFilteringProcessor.cs:18]: for my learning, why does the SpanId matter here?

Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/LogFilteringProcessor.cs:18]: logs that are not emitted within an activity context will match this criteria.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs [hunk 4]
```diff
@@ -138,7 +137,28 @@ public static OpenTelemetryBuilder UseAzureMonitor(this OpenTelemetryBuilder bui
...
+                        try
+                        {
+                            enableLogSamplingEnvVar = Environment.GetEnvironmentVariable(EnableLogSamplingEnvVar);
+                        }
+                        catch (Exception ex)
+                        {
+                            AzureMonitorAspNetCoreEventSource.Log.GetEnvironmentVariableFailed(EnableLogSamplingEnvVar, ex);
+                        }
+
+                        if (enableLogSamplingEnvVar != null && enableLogSamplingEnvVar.Equals("true", StringComparison.OrdinalIgnoreCase))
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs:152]: nit: we can avoid the null check if you move the boolean eval into the try at line 145.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/E2EAzureMonitorDistroTests.cs [hunk 5]
```diff
@@ -77,6 +80,44 @@ public async Task ValidateTelemetryExport()
...
+                    {
+                        Assert.True(processor is LogFilteringProcessor);
+                        Assert.True(processor is BatchLogRecordExportProcessor);
+                    }
+                    else
+                    {
+                        Assert.True(processor is not LogFilteringProcessor);
+                        Assert.True(processor is BatchLogRecordExportProcessor);
+                    }
+                }
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/E2EAzureMonitorDistroTests.cs:113]: I'm thinking we need more cases. 
`else if processor == null`

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/AzureMonitorAspNetCoreEventSource.cs [hunk 6]
```diff
@@ -62,5 +71,8 @@ public void ConfigureFailed(Exception ex)
 
         [Event(4, Message = "Vendor instrumentation added for: {0}.", Level = EventLevel.Verbose)]
         public void VendorInstrumentationAdded(string packageName) => WriteEvent(4, packageName);
+
+        [Event(5, Message = "Failed to Read environment variable {0}, exception: {1}", Level = EventLevel.Error)]
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/AzureMonitorAspNetCoreEventSource.cs:75]: nit: is this the first time this package is having to read an env variable?

**Comment by cijothomas** [comment]
> `OTEL_DOTNET_AZURE_MONITOR_EXPERIMENTAL_ENABLE_LOG_SAMPLING` when set to true will sample the logs that are emitted within the context of an activity. The sampling will done based on the SampleRate configured by user for traces.

Could you ensure the PR description is accurate? There is "SampleRate" sampling for logs.

**Comment by cijothomas** [comment]
@vishweshbankwar Is this behavior going to be same across other languages as well?

**Comment by vishweshbankwar** [comment]
> @vishweshbankwar Is this behavior going to be same across other languages as well?

Yes, that is the plan.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Azure.Monitor.OpenTelemetry.AspNetCore.csproj (hunk 1 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Azure.Monitor.OpenTelemetry.AspNetCore.csproj:30]: Done

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/E2EAzureMonitorDistroTests.cs (hunk 5 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/E2EAzureMonitorDistroTests.cs:113]: added assert

**Review by rajkumar-rangaraj — APPROVED** [review]
LGTM

**Review by TimothyMothra — APPROVED** [review]

**Comment by TimothyMothra** [comment]
CI failure:
> sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/tests/EventHubEndToEndTests.cs(889,29): Error CS1729: 'CheckpointPosition' does not contain a constructor that takes 2 arguments



@vishweshbankwar, the fix for this has already been approved and should merge soon. #41871
After this merges, lets merge main into your branch and rerun ci.
